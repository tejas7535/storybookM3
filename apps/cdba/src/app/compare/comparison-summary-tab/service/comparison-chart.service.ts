import { Injectable } from '@angular/core';

import { TranslocoService } from '@jsverse/transloco';

@Injectable({ providedIn: 'root' })
export class ComparisonChartService {
  constructor(private readonly transloco: TranslocoService) {}

  readonly PURCHASE = this.transloco.translate(
    'compare.summary.common.purchase'
  );
  readonly REMAINDER = this.transloco.translate(
    'compare.summary.common.remainder'
  );
  readonly MOH = this.transloco.translate('compare.summary.common.moh');
  readonly BURDEN = this.transloco.translate('compare.summary.common.burden');
  readonly LABOUR = this.transloco.translate('compare.summary.common.labour');
  readonly RAW_MATERIAL = this.transloco.translate(
    'compare.summary.common.rawMaterial'
  );
  readonly TOTAL = this.transloco.translate('compare.summary.common.total');
}
