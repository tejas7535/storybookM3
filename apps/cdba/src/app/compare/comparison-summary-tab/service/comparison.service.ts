import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { Observable } from 'rxjs';

import { API, ComparisonPath } from '@cdba/shared/constants/api';
import { BomIdentifier } from '@cdba/shared/models';

import { Comparison } from '../../../shared/models/comparison.model';

@Injectable({ providedIn: 'root' })
export class ComparisonService {
  constructor(private readonly httpClient: HttpClient) {}

  getComparison(identifiers: BomIdentifier[]): Observable<Comparison> {
    const path = `${API.v1}/${ComparisonPath}`;

    return this.httpClient.post<Comparison>(path, identifiers);
  }
}
