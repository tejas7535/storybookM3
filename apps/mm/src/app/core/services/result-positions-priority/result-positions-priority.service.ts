import { Injectable } from '@angular/core';

import {
  ResultItem,
  ResultItemWithTitle,
} from '@mm/core/store/models/calculation-result-state.model';

@Injectable({
  providedIn: 'root',
})
export class ResultPositionsPriorityService {
  private readonly startItemsImportanceAbbreviations = ['p_oil_ini', 'n_sld'];

  /** some data will be added here as mm 2.0 extend this and remove it once data will be implemented
   * - Control value bearing clearance class (order 2)
   * - Check values for clearance classes	(order 3)
   * - Radial clearance reduction (order 4)
   * based on the following: https://confluence.schaeffler.com/pages/viewpage.action?pageId=404420489
   * remove comment after implementation
   */
  private readonly endItemsImportanceAbbreviations = [
    'dx_mnt',
    'Fx_mnt',
    // 'Ïƒ_t_rcw',
    // 'p_jnt_mnt',
  ];

  getPrioritizedStartItems(positions: ResultItem[]): ResultItem[] {
    return this.prioritizeItems(
      positions,
      this.startItemsImportanceAbbreviations
    );
  }

  getPrioritizedEndItems(positions: ResultItem[]): ResultItem[] {
    return this.prioritizeItems(
      positions,
      this.endItemsImportanceAbbreviations
    );
  }

  getPrioritizedAndFormattedRadialClearance(
    radialClearances: ResultItemWithTitle[] | undefined
  ): ResultItem[] {
    if (!radialClearances) {
      return [];
    }

    const reversedClearances = this.reverseOrder(radialClearances);

    return this.mapResultItemWithTitleIntoResultItem(reversedClearances);
  }

  getPrioritizedClearanceClasses(
    clearanceClasses: ResultItemWithTitle[] | undefined
  ): ResultItem[] {
    if (!clearanceClasses) {
      return [];
    }

    return this.mapResultItemWithTitleIntoResultItem(clearanceClasses);
  }

  private mapResultItemWithTitleIntoResultItem(
    items: ResultItemWithTitle[]
  ): ResultItem[] {
    return items.flatMap((item) =>
      item.items.map((subItem) => ({
        ...subItem,
        designation: `${item.title} ( ${subItem.designation} )`,
        isImportant: true,
      }))
    );
  }

  private reverseOrder(items: ResultItemWithTitle[]): ResultItemWithTitle[] {
    return [...items].reverse();
  }

  private prioritizeItems(
    positions: ResultItem[],
    importanceList: string[]
  ): ResultItem[] {
    if (!positions?.length) {
      return [];
    }

    const prioritizedItems = positions.map((item) => ({
      ...item,
      isImportant: importanceList.includes(item.abbreviation),
    }));

    return this.sortByImportance(prioritizedItems, importanceList);
  }

  private sortByImportance(
    items: ResultItem[],
    importanceOrder: string[]
  ): ResultItem[] {
    return [...items].sort((a, b) => {
      if (a.isImportant && b.isImportant) {
        const indexA = importanceOrder.indexOf(a.abbreviation);
        const indexB = importanceOrder.indexOf(b.abbreviation);

        return indexA - indexB;
      }

      if (a.isImportant) {
        return -1;
      }

      if (b.isImportant) {
        return 1;
      }

      return 0;
    });
  }
}
