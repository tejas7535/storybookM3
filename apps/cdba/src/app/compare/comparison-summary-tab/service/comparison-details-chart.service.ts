import { Injectable } from '@angular/core';

import { TranslocoLocaleService } from '@jsverse/transloco-locale';
import { EChartsOption } from 'echarts';

import { DataPoint } from '@cdba/shared/components/bom-chart/data-point.model';
import { ComparisonType } from '@cdba/shared/constants';
import {
  CHART_HIGHLIGHT_ERROR,
  CHART_PRIMARY,
  CHART_SECONDARY,
  COST_SHARE_CATEGORY_COLORS,
} from '@cdba/shared/constants/colors';
import { ComparisonDetail } from '@cdba/shared/models/comparison.model';
import { CostShareService } from '@cdba/shared/services';

import { ComparisonChartService } from './comparison-chart.service';

@Injectable({ providedIn: 'root' })
export class ComparisonDetailsChartService {
  constructor(
    private readonly comparisonChartService: ComparisonChartService,
    private readonly localeService: TranslocoLocaleService,
    private readonly costShareService: CostShareService
  ) {}

  provideParetoChartConfig(
    details: ComparisonDetail[],
    currency: string
  ): EChartsOption {
    const xAxisData: string[] = [];
    const yAxisLineData: number[] = [];
    const yAxisBarDataPoints: DataPoint[] = [];
    let min = Number.POSITIVE_INFINITY;
    let max = Number.NEGATIVE_INFINITY;
    let cumulativePercentageCost = 0;

    details.forEach((detail) => {
      if (detail.costSharePercentage !== 0) {
        xAxisData.push(detail.title);
        min = Math.min(min, detail.totalCosts);
        max = Math.max(max, detail.totalCosts);

        cumulativePercentageCost += detail.costSharePercentage;
        cumulativePercentageCost = Number.parseFloat(
          cumulativePercentageCost.toPrecision(4)
        );
        yAxisLineData.push(cumulativePercentageCost);
        yAxisBarDataPoints.push({
          name: detail.title,
          value: detail.totalCosts,
          itemStyle: {
            color: COST_SHARE_CATEGORY_COLORS.get(
              this.costShareService.getCostShareCategory(
                detail.costSharePercentage / 100
              )
            ),
          },
        });
      }
    });

    return {
      tooltip: {
        trigger: 'axis',
        axisPointer: {
          type: 'cross',
          crossStyle: {
            color: '#999',
          },
        },
      },
      toolbox: {
        feature: {
          saveAsImage: { show: true },
        },
      },
      xAxis: [
        {
          type: 'category',
          data: xAxisData,
          axisPointer: {
            type: 'shadow',
          },
        },
      ],
      yAxis: [
        {
          type: 'value',
          min,
          max,
          axisLabel: {
            formatter: (value: string) => `${value} ${currency}`,
          },
        },
        {
          type: 'value',
          min: 0,
          max: 100,
          interval: 10,
          axisLabel: {
            formatter: (value: string) => `${value} %`,
          },
        },
      ],
      series: [
        {
          type: 'bar',
          tooltip: {
            valueFormatter: (value: number) =>
              `${this.localeService.localizeNumber(value, 'decimal')} ${currency}`,
          },
          data: yAxisBarDataPoints,
        },
        {
          type: 'line',
          color: CHART_HIGHLIGHT_ERROR,
          yAxisIndex: 1,
          tooltip: {
            valueFormatter: (value: string) => `${value} %`,
          },
          data: yAxisLineData,
        },
      ],
    } as EChartsOption;
  }

  provideDetailedBarChartConfig(
    detail: ComparisonDetail,
    firstMaterialDesignation: string,
    secondMaterialDesignation: string
  ): EChartsOption {
    const firstMaterialDesignationCosts: number[] = [];
    const secondMaterialDesignationCosts: number[] = [];

    detail.costDifferences.forEach((costDifference) => {
      switch (costDifference.type) {
        case ComparisonType.PURCHASE: {
          firstMaterialDesignationCosts[0] = costDifference.valueBom1;
          secondMaterialDesignationCosts[0] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.RAW_MATERIAL: {
          firstMaterialDesignationCosts[1] = costDifference.valueBom1;
          secondMaterialDesignationCosts[1] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.LABOUR: {
          firstMaterialDesignationCosts[2] = costDifference.valueBom1;
          secondMaterialDesignationCosts[2] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.BURDEN: {
          firstMaterialDesignationCosts[3] = costDifference.valueBom1;
          secondMaterialDesignationCosts[3] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.MOH: {
          firstMaterialDesignationCosts[4] = costDifference.valueBom1;
          secondMaterialDesignationCosts[4] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.REMAINDER: {
          firstMaterialDesignationCosts[5] = costDifference.valueBom1;
          secondMaterialDesignationCosts[5] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.TOTAL: {
          firstMaterialDesignationCosts[6] = costDifference.valueBom1;
          secondMaterialDesignationCosts[6] = costDifference.valueBom2;
          break;
        }
        default: {
          throw new Error(`Invalid Comparison Type: ${costDifference.type}`);
        }
      }
    });

    return {
      tooltip: {
        trigger: 'axis',
        axisPointer: {
          type: 'shadow',
        },
      },
      xAxis: {
        type: 'category',
        data: [
          this.comparisonChartService.PURCHASE,
          this.comparisonChartService.RAW_MATERIAL,
          this.comparisonChartService.LABOUR,
          this.comparisonChartService.BURDEN,
          this.comparisonChartService.MOH,
          this.comparisonChartService.REMAINDER,
          this.comparisonChartService.TOTAL,
        ],
        axisLabel: {
          interval: 0,
          width: 200,
        },
      },
      yAxis: {
        type: 'value',
      },
      series: [
        {
          type: 'bar',
          name: firstMaterialDesignation,
          data: firstMaterialDesignationCosts,
          itemStyle: {
            color: CHART_PRIMARY,
          },
        },
        {
          type: 'bar',
          name: secondMaterialDesignation,
          data: secondMaterialDesignationCosts,
          itemStyle: {
            color: CHART_SECONDARY,
          },
        },
      ],
    } as EChartsOption;
  }
}
