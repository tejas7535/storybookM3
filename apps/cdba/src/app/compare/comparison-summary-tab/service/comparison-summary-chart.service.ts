import { Injectable } from '@angular/core';

import { TranslocoService } from '@jsverse/transloco';
import { EChartsOption } from 'echarts/types/dist/echarts';

import {
  CHART_PRIMARY,
  CHART_SECONDARY,
  RADAR_CHART_REFERENCE_LINE,
} from '@cdba/shared/constants/colors';
import { ComparisonType } from '@cdba/shared/constants/comparison-type';
import {
  Comparison,
  ComparisonSummary,
  CostDifference,
} from '@cdba/shared/models/comparison.model';

import { ComparisonChartService } from './comparison-chart.service';

@Injectable({ providedIn: 'root' })
export class ComparisonSummaryChartService {
  constructor(
    private readonly transloco: TranslocoService,
    private readonly comparisonChartService: ComparisonChartService
  ) {}

  provideSummaryTableColumns(): string[] {
    return [
      'costType',
      'valueBom1',
      'valueBom2',
      'currency',
      'costDeltaValue',
      'costDeltaPercentage',
    ];
  }

  provideSummarizedRadarChartConfig(
    firstMaterialDesignation: string,
    secondMaterialDesignation: string,
    comparison: Comparison
  ): EChartsOption {
    const referenceLineTitle = this.transloco.translate(
      'compare.summary.common.referenceLine'
    );

    const purchaseCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.PURCHASE
    );
    const remainderCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.REMAINDER
    );
    const mohCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.MOH
    );
    const burdenCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.BURDEN
    );
    const labourCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.LABOUR
    );
    const rawMaterialCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.RAW_MATERIAL
    );

    // Calculate min max values
    let min = Number.POSITIVE_INFINITY;
    let max = Number.NEGATIVE_INFINITY;
    comparison.summary.costDifferences.forEach((costDifference) => {
      if (costDifference.type !== 'TOTAL') {
        min = Math.min(min, costDifference.valueBom1, costDifference.valueBom2);
        max = Math.max(max, costDifference.valueBom1, costDifference.valueBom2);
      }
    });
    min = Math.floor(min);
    max = Math.ceil(max);

    return {
      legend: {
        bottom: '0%',
        left: '5%',
        data: [
          referenceLineTitle,
          firstMaterialDesignation,
          secondMaterialDesignation,
        ],
      },
      tooltip: {
        trigger: 'item',
        valueFormatter: (value: string) => `${value} ${comparison.currency}`,
      },
      toolbox: {
        feature: {
          saveAsImage: { show: true },
        },
      },
      radar: {
        indicator: [
          {
            name: this.comparisonChartService.PURCHASE,
            min,
            max,
          },
          {
            name: this.comparisonChartService.REMAINDER,
            min,
            max,
          },
          {
            name: this.comparisonChartService.MOH,
            min,
            max,
          },
          {
            name: this.comparisonChartService.BURDEN,
            min,
            max,
          },
          {
            name: this.comparisonChartService.LABOUR,
            min,
            max,
          },
          {
            name: this.comparisonChartService.RAW_MATERIAL,
            min,
            max,
          },
        ],
      },
      series: [
        {
          type: 'radar',
          data: [
            {
              name: referenceLineTitle,
              value: [0, 0, 0, 0, 0, 0],
              itemStyle: {
                color: RADAR_CHART_REFERENCE_LINE,
              },
            },
            {
              name: firstMaterialDesignation,
              value: [
                purchaseCost.valueBom1,
                remainderCost.valueBom1,
                mohCost.valueBom1,
                burdenCost.valueBom1,
                labourCost.valueBom1,
                rawMaterialCost.valueBom1,
              ],
              itemStyle: {
                color: CHART_PRIMARY,
              },
            },
            {
              name: secondMaterialDesignation,
              value: [
                purchaseCost.valueBom2,
                remainderCost.valueBom2,
                mohCost.valueBom2,
                burdenCost.valueBom2,
                labourCost.valueBom2,
                rawMaterialCost.valueBom2,
              ],
              itemStyle: {
                color: CHART_SECONDARY,
              },
            },
          ],
        },
      ],
    } as EChartsOption;
  }

  findCostDifferenceByComparisonType(
    summary: ComparisonSummary,
    type: ComparisonType
  ): CostDifference {
    return summary.costDifferences.find(
      (costDifference) => costDifference.type === type
    );
  }
}
