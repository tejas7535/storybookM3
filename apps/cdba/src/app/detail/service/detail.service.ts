import { HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { DataService } from '../../core/http/data.service';
import {
  BomIdentifier,
  BomItem,
  BomResult,
  ReferenceTypeIdModel,
  ReferenceTypeResultModel,
} from '../../core/store/reducers/detail/models';
import { CalculationsResultModel } from '../../core/store/reducers/detail/models/calculations-result-model';
import { Calculation } from '../../core/store/reducers/shared/models/calculation.model';

@Injectable({
  providedIn: 'root',
})
export class DetailService {
  private readonly DETAIL = 'detail';

  private readonly CALCULATIONS = 'calculations';

  private readonly PARAM_MATERIAL_NUMBER = 'material-number';
  private readonly PARAM_PLANT = 'plant';

  private readonly BOM = 'bom';
  private readonly PARAM_BOM_COSTING_DATE = 'bom-costing-date';
  private readonly PARAM_BOM_COSTING_NUMBER = 'bom-costing-number';
  private readonly PARAM_BOM_COSTING_TYPE = 'bom-costing-type';
  private readonly PARAM_BOM_COSTING_VERSION = 'bom-costing-version';
  private readonly PARAM_BOM_ENTERED_MANUALLY = 'bom-entered-manually';
  private readonly PARAM_BOM_REFERENCE_OBJECT = 'bom-reference-object';
  private readonly PARAM_BOM_VALUATION_VARIANT = 'bom-valuation-variant';

  private static defineBomTreeForAgGrid(
    items: BomItem[],
    idx: number
  ): BomItem[] {
    if (idx === items.length) {
      return items;
    }

    const currentItem = items[idx];
    const nextIdx = idx + 1;

    // special case level 1
    if (currentItem.level === 1) {
      currentItem.predecessorsInTree = [currentItem.materialDesignation];

      return DetailService.defineBomTreeForAgGrid(items, nextIdx);
    }

    const previousItem = items[idx - 1];

    if (previousItem.level < currentItem.level) {
      // previous item is parent
      currentItem.predecessorsInTree = [
        ...previousItem.predecessorsInTree,
        currentItem.materialDesignation,
      ];
    } else {
      // previous item has same parent or different parent
      const itemsToRemove = previousItem.level - currentItem.level + 1;
      const predecessors = [...previousItem.predecessorsInTree];

      // remove last items to have the actual parent as last item
      predecessors.splice(predecessors.length - itemsToRemove, itemsToRemove);

      // since duplicates are possible, add spaces to the key until it is unique
      let foundUniqueKey = false;
      let currentMaterialDesignation = currentItem.materialDesignation;
      while (!foundUniqueKey) {
        const temp = [...predecessors, currentMaterialDesignation];

        if (
          items
            .map((item) => item.predecessorsInTree)
            .find(
              (materialDesignations: string[]) =>
                materialDesignations.length === temp.length &&
                materialDesignations.every(
                  (value: string, index: number) => value === temp[index]
                )
            ) !== undefined
        ) {
          currentMaterialDesignation += ' ';
        } else {
          foundUniqueKey = true;
        }
      }

      currentItem.predecessorsInTree = [
        ...predecessors,
        currentMaterialDesignation,
      ];
    }

    return DetailService.defineBomTreeForAgGrid(items, nextIdx);
  }

  public constructor(private readonly dataService: DataService) {}

  public detail(
    item: ReferenceTypeIdModel
  ): Observable<ReferenceTypeResultModel> {
    const params: HttpParams = new HttpParams()
      .set(this.PARAM_MATERIAL_NUMBER, item.materialNumber)
      .set(this.PARAM_PLANT, item.plant);

    return this.dataService.getAll<ReferenceTypeResultModel>(
      this.DETAIL,
      params
    );
  }

  public calculations(materialNumber: string): Observable<Calculation[]> {
    const params = new HttpParams().set(
      this.PARAM_MATERIAL_NUMBER,
      materialNumber
    );

    return this.dataService
      .getAll<CalculationsResultModel>(this.CALCULATIONS, params)
      .pipe(map((response: CalculationsResultModel) => response.items));
  }

  public getBom(bomIdentifier: BomIdentifier): Observable<BomItem[]> {
    const params = new HttpParams()
      .set(this.PARAM_BOM_COSTING_DATE, bomIdentifier.bomCostingDate)
      .set(this.PARAM_BOM_COSTING_NUMBER, bomIdentifier.bomCostingNumber)
      .set(this.PARAM_BOM_COSTING_TYPE, bomIdentifier.bomCostingType)
      .set(this.PARAM_BOM_COSTING_VERSION, bomIdentifier.bomCostingVersion)
      .set(this.PARAM_BOM_ENTERED_MANUALLY, bomIdentifier.bomEnteredManually)
      .set(this.PARAM_BOM_REFERENCE_OBJECT, bomIdentifier.bomReferenceObject)
      .set(this.PARAM_BOM_VALUATION_VARIANT, bomIdentifier.bomValuationVariant);

    return this.dataService.getAll<BomResult>(this.BOM, params).pipe(
      map((response: BomResult) =>
        response.items.map((item) => ({
          ...item,
          predecessorsInTree: [],
        }))
      ),
      map((items: BomItem[]) => DetailService.defineBomTreeForAgGrid(items, 0))
    );
  }
}
