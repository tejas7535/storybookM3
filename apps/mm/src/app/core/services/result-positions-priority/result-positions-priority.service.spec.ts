/* eslint-disable unicorn/no-useless-undefined */
import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';

import {
  ALL_IMPORTANT_END_POSITIONS,
  ALL_IMPORTANT_POSITIONS,
  CLEARANCE_CLASSES_RESULT_WITH_ITEMS,
  EMPTY_END_POSITIONS,
  EMPTY_POSITIONS,
  MIXED_IMPORTANT_END_POSITIONAL_REVERSE_ORDER,
  MIXED_IMPORTANT_POSITIONAL_REVERSE_ORDER,
  NO_IMPORTANT_END_POSITIONS,
  NO_IMPORTANT_POSITIONS,
  RADIAL_CLEARANCE_RESULT_WITH_ITEMS,
  SOME_IMPORTANT_END_POSITIONS,
  SOME_IMPORTANT_POSITIONS,
} from './../../../../testing/mocks/result-positions-priority.mock';
import { ResultPositionsPriorityService } from './result-positions-priority.service';

describe('ResultPositionsPriorityService', () => {
  let spectator: SpectatorService<ResultPositionsPriorityService>;
  let service: ResultPositionsPriorityService;

  const createService = createServiceFactory({
    service: ResultPositionsPriorityService,
  });

  beforeEach(() => {
    spectator = createService();
    service = spectator.service;
  });

  it('should be created', () => {
    expect(service).toBeTruthy();
  });

  describe('getPrioritizedStartItems', () => {
    it('should return empty array when input is empty', () => {
      const result = service.getPrioritizedStartItems(EMPTY_POSITIONS);
      expect(result).toEqual([]);
    });

    it('should mark no items as important when none match important abbreviations', () => {
      const result = service.getPrioritizedStartItems(NO_IMPORTANT_POSITIONS);
      expect(result.some((item) => item.isImportant)).toBeFalsy();
      expect(result).toMatchSnapshot();
    });

    it('should mark items as important and put them first in result', () => {
      const result = service.getPrioritizedStartItems(SOME_IMPORTANT_POSITIONS);
      // Keep one critical assertion explicit
      expect(result[0].abbreviation).toBe('p_oil_ini');
      expect(result).toMatchSnapshot();
    });

    it('should preserve all items in the result', () => {
      const result = service.getPrioritizedStartItems(SOME_IMPORTANT_POSITIONS);
      expect(result).toHaveLength(SOME_IMPORTANT_POSITIONS.length);
    });

    it('should sort important items according to the predefined order', () => {
      const result = service.getPrioritizedStartItems(ALL_IMPORTANT_POSITIONS);
      expect(result).toMatchSnapshot();
    });

    it('should handle mixed important items and place them in correct order', () => {
      const result = service.getPrioritizedStartItems(
        MIXED_IMPORTANT_POSITIONAL_REVERSE_ORDER
      );
      expect(result).toMatchSnapshot();
    });
  });

  describe('getPrioritizedEndItems', () => {
    it('should return empty array when input is empty', () => {
      const result = service.getPrioritizedEndItems(EMPTY_END_POSITIONS);
      expect(result).toEqual([]);
    });

    it('should mark no items as important when none match important abbreviations', () => {
      const result = service.getPrioritizedEndItems(NO_IMPORTANT_END_POSITIONS);
      expect(result.some((item) => item.isImportant)).toBeFalsy();
      expect(result).toMatchSnapshot();
    });

    it('should mark items as important and put them first in result', () => {
      const result = service.getPrioritizedEndItems(
        SOME_IMPORTANT_END_POSITIONS
      );
      // Check if important items are at the beginning
      expect(result[0].abbreviation).toBe('dx_mnt');
      expect(result).toMatchSnapshot();
    });

    it('should preserve all items in the result', () => {
      const result = service.getPrioritizedEndItems(
        SOME_IMPORTANT_END_POSITIONS
      );
      expect(result).toHaveLength(SOME_IMPORTANT_END_POSITIONS.length);
    });

    it('should sort important items according to the predefined order', () => {
      const result = service.getPrioritizedEndItems(
        ALL_IMPORTANT_END_POSITIONS
      );

      // Verify order: 'dx_mnt' should come before 'Fx_mnt'
      expect(result[0].abbreviation).toBe('dx_mnt');
      expect(result[1].abbreviation).toBe('Fx_mnt');
      expect(result).toMatchSnapshot();
    });

    it('should handle mixed important items and place them in correct order', () => {
      const result = service.getPrioritizedEndItems(
        MIXED_IMPORTANT_END_POSITIONAL_REVERSE_ORDER
      );
      expect(result).toMatchSnapshot();
    });
  });

  describe('getPrioritizedAndFormattedRadialClearance', () => {
    it('should return empty array when input is undefined', () => {
      const result =
        service.getPrioritizedAndFormattedRadialClearance(undefined);
      expect(result).toEqual([]);
    });

    it('should correctly transform radial clearance data', () => {
      const result = service.getPrioritizedAndFormattedRadialClearance(
        RADIAL_CLEARANCE_RESULT_WITH_ITEMS
      );

      expect(result.length).toBe(3);
      expect(result[0].designation).toBe('Clearance 2 ( Item 3 )');
      expect(result.every((item) => item.isImportant)).toBe(true);

      expect(result[0].abbreviation).toBe('c2i1');

      expect(result).toMatchSnapshot();
    });
  });

  describe('getPrioritizedClearanceClasses', () => {
    it('should return empty array when input is undefined', () => {
      const result = service.getPrioritizedClearanceClasses(undefined);
      expect(result).toEqual([]);
    });

    it('should transform data correctly', () => {
      const result = service.getPrioritizedClearanceClasses(
        CLEARANCE_CLASSES_RESULT_WITH_ITEMS
      );
      expect(result).toMatchSnapshot();
    });
  });
});
