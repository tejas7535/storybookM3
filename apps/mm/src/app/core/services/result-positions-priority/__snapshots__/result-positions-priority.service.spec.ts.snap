// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultPositionsPriorityService getPrioritizedAndFormattedRadialClearance should correctly transform radial clearance data 1`] = `
Array [
  Object {
    "abbreviation": "c2i1",
    "designation": "Clearance 2 ( Item 3 )",
    "isImportant": true,
    "title": "Original Title 3",
    "value": "0.3 mm",
  },
  Object {
    "abbreviation": "c1i1",
    "designation": "Clearance 1 ( Item 1 )",
    "isImportant": true,
    "value": "0.1 mm",
  },
  Object {
    "abbreviation": "c1i2",
    "designation": "Clearance 1 ( Item 2 )",
    "isImportant": true,
    "value": "0.2 mm",
  },
]
`;

exports[`ResultPositionsPriorityService getPrioritizedClearanceClasses should transform data correctly 1`] = `
Array [
  Object {
    "abbreviation": "c1i1",
    "designation": "Class 1 ( Item 1 )",
    "isImportant": true,
    "value": "0.1 mm",
  },
  Object {
    "abbreviation": "c1i2",
    "designation": "Class 1 ( Item 2 )",
    "isImportant": true,
    "value": "0.2 mm",
  },
  Object {
    "abbreviation": "c2i1",
    "designation": "Class 2 ( Item 3 )",
    "isImportant": true,
    "value": "0.3 mm",
  },
]
`;

exports[`ResultPositionsPriorityService getPrioritizedEndItems should handle mixed important items and place them in correct order 1`] = `
Array [
  Object {
    "abbreviation": "dx_mnt",
    "designation": "Displacement (start to end position)",
    "isImportant": true,
    "unit": "mm",
    "value": "5.397",
  },
  Object {
    "abbreviation": "Fx_mnt",
    "designation": "Required mounting load",
    "isImportant": true,
    "unit": "kN",
    "value": "1482.9",
  },
  Object {
    "abbreviation": "p_jnt_mnt",
    "designation": "Joint pressure after mounting",
    "isImportant": false,
    "unit": "N/mm²",
    "value": "10.7",
  },
  Object {
    "abbreviation": "σ_t_rcw",
    "designation": "Tangential stress (raceway)",
    "isImportant": false,
    "unit": "N/mm²",
    "value": "94.1",
  },
]
`;

exports[`ResultPositionsPriorityService getPrioritizedEndItems should mark items as important and put them first in result 1`] = `
Array [
  Object {
    "abbreviation": "dx_mnt",
    "designation": "Displacement (start to end position)",
    "isImportant": true,
    "unit": "mm",
    "value": "5.397",
  },
  Object {
    "abbreviation": "p_jnt_mnt",
    "designation": "Joint pressure after mounting",
    "isImportant": false,
    "unit": "N/mm²",
    "value": "10.7",
  },
  Object {
    "abbreviation": "σ_t_b",
    "designation": "Tangential stress (bore)",
    "isImportant": false,
    "unit": "N/mm²",
    "value": "103.9",
  },
]
`;

exports[`ResultPositionsPriorityService getPrioritizedEndItems should mark no items as important when none match important abbreviations 1`] = `
Array [
  Object {
    "abbreviation": "eps_d_ir",
    "designation": "Inner raceway expansion",
    "isImportant": false,
    "unit": "‰",
    "value": "0.450",
  },
  Object {
    "abbreviation": "Fx_hld",
    "designation": "Holding load",
    "isImportant": false,
    "unit": "kN",
    "value": "610.6",
  },
]
`;

exports[`ResultPositionsPriorityService getPrioritizedEndItems should sort important items according to the predefined order 1`] = `
Array [
  Object {
    "abbreviation": "dx_mnt",
    "designation": "Displacement (start to end position)",
    "isImportant": true,
    "unit": "mm",
    "value": "5.397",
  },
  Object {
    "abbreviation": "Fx_mnt",
    "designation": "Required mounting load",
    "isImportant": true,
    "unit": "kN",
    "value": "1482.9",
  },
]
`;

exports[`ResultPositionsPriorityService getPrioritizedStartItems should handle mixed important items and place them in correct order 1`] = `
Array [
  Object {
    "abbreviation": "p_oil_ini",
    "designation": "Required initial oil pressure",
    "isImportant": true,
    "unit": "bar",
    "value": "19.88",
  },
  Object {
    "abbreviation": "n_sld",
    "designation": "Number of sliding surfaces",
    "isImportant": true,
    "unit": "-",
    "value": "1",
  },
  Object {
    "abbreviation": "p_jnt_ini",
    "designation": "Initial joint pressure",
    "isImportant": false,
    "unit": "N/mm²",
    "value": "2.1",
  },
  Object {
    "abbreviation": "Fx_ini",
    "designation": "Required initial load",
    "isImportant": false,
    "unit": "kN",
    "value": "296.6",
  },
]
`;

exports[`ResultPositionsPriorityService getPrioritizedStartItems should mark items as important and put them first in result 1`] = `
Array [
  Object {
    "abbreviation": "p_oil_ini",
    "designation": "Required initial oil pressure",
    "isImportant": true,
    "unit": "bar",
    "value": "19.88",
  },
  Object {
    "abbreviation": "Fx_ini",
    "designation": "Required initial load",
    "isImportant": false,
    "unit": "kN",
    "value": "296.6",
  },
  Object {
    "abbreviation": "p_jnt_ini",
    "designation": "Initial joint pressure",
    "isImportant": false,
    "unit": "N/mm²",
    "value": "2.1",
  },
]
`;

exports[`ResultPositionsPriorityService getPrioritizedStartItems should mark no items as important when none match important abbreviations 1`] = `
Array [
  Object {
    "abbreviation": "Fx_ini",
    "designation": "Required initial load",
    "isImportant": false,
    "unit": "kN",
    "value": "296.6",
  },
  Object {
    "abbreviation": "p_jnt_ini",
    "designation": "Initial joint pressure",
    "isImportant": false,
    "unit": "N/mm²",
    "value": "2.1",
  },
]
`;

exports[`ResultPositionsPriorityService getPrioritizedStartItems should sort important items according to the predefined order 1`] = `
Array [
  Object {
    "abbreviation": "p_oil_ini",
    "designation": "Required initial oil pressure",
    "isImportant": true,
    "unit": "bar",
    "value": "19.88",
  },
  Object {
    "abbreviation": "n_sld",
    "designation": "Number of sliding surfaces",
    "isImportant": true,
    "unit": "-",
    "value": "1",
  },
]
`;
